pkg/noderesourcetopology/least_numa.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package noderesourcetopology
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+	v1qos "k8s.io/kubernetes/pkg/apis/core/v1/helper/qos"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	topologyv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+	"gonum.org/v1/gonum/stat/combin"
+	"sigs.k8s.io/scheduler-plugins/pkg/util"
+)
+
+var zeroQty resource.Quantity
+
+func init() {
+	zeroQty = resource.MustParse("0")
+}
+
+func leastNUMAContainerScopeScore(pod *v1.Pod, zones topologyv1alpha1.ZoneList) (int64, *framework.Status) {
+	nodes := createNUMANodeList(zones)
+	qos := v1qos.GetPodQOS(pod)
+
+	maxNuma := 0
+	// the order how TopologyManager asks for hint is important so doing it in the same order
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/scope_container.go#L52
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		identifier := fmt.Sprintf("%s/%s/%s", pod.Namespace, pod.Name, container.Name)
+		numaNodes, numaNodesCount := numaNodesRequired(identifier, qos, nodes, container.Resources.Requests)
+		if numaNodesCount < 1 {
+			errMsg := fmt.Sprintf("cannot calculate NUMA score for container: %s", container.Name)
+			klog.ErrorS(fmt.Errorf(errMsg), "failed to calculate numaNodesRequired for: %s", identifier)
+			return 0, framework.NewStatus(framework.Error, errMsg)
+		}
+
+		if numaNodesCount > maxNuma {
+			maxNuma = numaNodesCount
+		}
+
```

Comment:
Just wondering if we need to handle the case of a very large pods that requires resources from more than 8 numa nodes assuming there are nodes with that capability in the cluster. Maybe we should mention here as being out of scope because  the maximum number of numa nodes supported by Topology manager is 8.

Reply:
I think the name of the variable is confusing. The intend of  this check was to see which container in a POD requires the most NUMA nodes and return it :) I'll change the variable name to `maxNUMANodesCount`

Reply:
renamed